[PDI-13408] - Metastore performance degrades with each object added

AndreyBurikhin opened the pull request:
@mdamour1976 please review.

These changes include fixed XmlMetaStore which is able to work with cache (to improve perfomance of add/get operations).
By default XmlMetaStore will not use cache. The developer have to add reference on cache's implementation as constructor's argument in XmlMetaStore, to add cache for kettle's XmlMetastore and other projects.

mdamour1976 commented:
@AndreyBurikhin Any chance you can squash these into a single commit?

AndreyBurikhin commented:
@mdamour1976  I've squashed them.

mdamour1976 commented on test-src/org/pentaho/metastore/stores/xml/AutomaticXmlMetaStoreCacheTest.java:
```diff
@@ -0,0 +1,26 @@
+public class AutomaticXmlMetaStoreCacheTest extends BaseXmlMetaStoreCacheTest {
+
+  @Override
+  protected XmlMetaStoreCache createMetaStore() {
```
Probably should rename this method to createMetaStoreCache()

mdamour1976 commented on test-src/org/pentaho/metastore/stores/xml/PersistentXmlMetaStoreCacheTest.java:
```diff
@@ -0,0 +1,25 @@
+
+  @Override
+  protected XmlMetaStoreCache createMetaStore() {
+    return new PersistentXmlMetaStoreCache();
```
and this one

mdamour1976 commented:
Looks good Andrey - only minor change.  I'm still debating if we should use automatic/persistent cache as default instead of none.  I will think about this.

AndreyBurikhin commented:
@mdamour1976 

I've renamed methods in AutomaticXmlMetaStoreCacheTest and PersistentXmlMetaStoreCacheTest.

As for default "no caching" strategy:  I've implemented it this way for one reason which is important for production systems - there are no unexpected changes for possible clients. If we modify  kettle or other client of xml metastore to use cache, we'll expect some changes of its performance, but we'll also expect drawbacks - it'll use more RAM (or even affects on GC).

However other products continue working with no sideeffects. 
We can hint other developers to swith on cache using metastore documentation. We also should deprecate XmlMetastre constructors without cache argument. 
So when their support teams will face with metastore performance issues, they will read docs (and javadocs) and found the recomendation to use cache.

...and it will be perfect if we can add Java library to this project. It'll allow us to use map with 'soft reference' values from the box
